### PR TITLE
Add react-leaflet

### PR DESCRIPTION
**Pull request by danzel**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `npm run new-package package-name`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

### CODE REVIEW

**Comment by danzel** [comment]
Looks like broken build isn't me, was broken in 8928fc6a3d6e0b7808b12bb4097c232015c7f865

File: react-leaflet/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,258 @@
+// Type definitions for react-leaflet 1.0
+// Project: https://github.com/PaulLeCam/react-leaflet
+// Definitions by: Dave Leaver <https://github.com/danzel>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
```
Comment (andy-ms) [inline react-leaflet/index.d.ts]: Add `// TypeScript Version: 2.1` right here (line 5) to fix tests.

**Comment by danzel** [comment]
Thanks, new errors are my fault.

**Comment by danzel** [comment]
Okay, build fixed! Sorry for the noise.

File: react-leaflet/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+//Path is an abstract class
+interface PathProps extends LeafletLayerEvents, LeafletInteractionEvents, Leaflet.PathOptions {
+}
+
+
+interface CircleProps extends PathProps {
+    center: Leaflet.LatLngExpression;
+    radius?: number;
+}
+interface Circle extends React.ComponentClass<CircleProps> { }
```
Comment (andy-ms) [inline react-leaflet/index.d.ts]: Why not just `declare const Circle: React.ComponentClass<CircleProps>`?

File: react-leaflet/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+// TypeScript Version: 2.1
+
+//tslint:disable:no-empty-interface
+
+/// <reference types="geojson" />
+
+import * as Leaflet from 'leaflet';
+import * as React from 'react';
+
+//All events need to be lowercase so they don't collide with React.DOMAttributes<T>
```
Comment (andy-ms) [inline react-leaflet/index.d.ts]: If the event names aren't accurate to the names the library actually uses, there's no point...

File: react-leaflet/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+}
+interface LayersControl extends React.ComponentClass<LayersControlProps> { }
+declare const LayersControl: LayersControl & { BaseLayer: LayersControl.BaseLayer, Overlay: LayersControl.Overlay };
+
+declare namespace LayersControl {
+    interface LayersControlLayerProps {
+        name: string;
+        checked?: boolean;
+    }
+    interface BaseLayer extends React.ComponentClass<LayersControlLayerProps> { }
```
Comment (andy-ms) [inline react-leaflet/index.d.ts]: Why not just `type BaseLayer = React.ComponentClass<LayersControlLayerProps>`?

File: react-leaflet/index.d.ts (hunk 3 above)
Comment (danzel) [inline react-leaflet/index.d.ts]: react-leaflet applies `toLowerCase` to all of the `on...` methods before using them for the internal leaflet library, so this works correctly. Both `onclick` and `onCLiCk` would work.
I've manually verified this as well.

https://github.com/PaulLeCam/react-leaflet/blob/d7ba0947b8fda95c1cdc54f1c5022cfd38306647/src/MapComponent.js#L42

Comment (andy-ms) [inline react-leaflet/index.d.ts]: Ah, OK. Comment made it sound like you changed it just for the sake of typings.

File: react-leaflet/index.d.ts (hunk 4 above)
Comment (danzel) [inline react-leaflet/index.d.ts]: thanks! didn't think of that, done now

File: react-leaflet/index.d.ts (hunk 2 above)
Comment (danzel) [inline react-leaflet/index.d.ts]: I had started on this and reverted it as I was running in to trouble with `no-empty-interface`.
Have re-done it now and used a few `tslint:disable-next-line:no-empty-interface` as I'm not sure of a better way to do some of these.
The big one is `PathProps`, which saves me putting those 3 interfaces on the 5 `*Props` interfaces below.
If you know of a nicer way to do this I'd gladly change it. Thanks :)

**Comment by alejo90** [comment]
👍

**Comment by PCaponetti** [comment]
@danzel do you see any issues with your typings and typescript 2.3+?  I'm getting issues where things are not assignable to `Readonly<P>` and have had to downgrade typescript to 2.2 to get it to compile.

**Comment by danzel** [comment]
I have not upgraded to typescript 2.3 yet, I think #16246 should be fixing it, but it has stalled

**Pull request merged**